File: backend/app/models/strength.py
"""Pydantic Schemas für Krafttraining API."""

from datetime import date
from enum import Enum
from typing import Optional

from pydantic import BaseModel, Field, model_validator


class ExerciseCategory(str, Enum):
    """Übungskategorie."""

    PUSH = "push"
    PULL = "pull"
    LEGS = "legs"
    CORE = "core"
    CARDIO = "cardio"
    DRILLS = "drills"


class SetStatus(str, Enum):
    """Status eines Satzes."""

    COMPLETED = "completed"
    REDUCED = "reduced"
    SKIPPED = "skipped"


class SetType(str, Enum):
    """Typ eines Satzes — bestimmt welche Felder relevant sind."""

    WEIGHT_REPS = "weight_reps"
    BODYWEIGHT_REPS = "bodyweight_reps"
    WEIGHTED_BODYWEIGHT = "weighted_bodyweight"
    ASSISTED_BODYWEIGHT = "assisted_bodyweight"
    DURATION = "duration"
    WEIGHT_DURATION = "weight_duration"
    DISTANCE_DURATION = "distance_duration"
    WEIGHT_DISTANCE = "weight_distance"


# Felder die je Set-Typ erforderlich bzw. erlaubt sind
SET_TYPE_FIELDS: dict[SetType, dict[str, list[str]]] = {
    SetType.WEIGHT_REPS: {"required": ["reps", "weight_kg"], "optional": []},
    SetType.BODYWEIGHT_REPS: {"required": ["reps"], "optional": []},
    SetType.WEIGHTED_BODYWEIGHT: {"required": ["reps", "weight_kg"], "optional": []},
    SetType.ASSISTED_BODYWEIGHT: {"required": ["reps", "weight_kg"], "optional": []},
    SetType.DURATION: {"required": ["duration_sec"], "optional": []},
    SetType.WEIGHT_DURATION: {"required": ["weight_kg", "duration_sec"], "optional": []},
    SetType.DISTANCE_DURATION: {"required": ["distance_m"], "optional": ["duration_sec"]},
    SetType.WEIGHT_DISTANCE: {"required": ["weight_kg", "distance_m"], "optional": []},
}


# --- Request Models ---


class SetInput(BaseModel):
    """Einzelner Satz einer Übung."""

    type: SetType = SetType.WEIGHT_REPS
    reps: int | None = Field(default=None, ge=0, le=999)
    weight_kg: float | None = Field(default=None, ge=0, le=999)
    duration_sec: int | None = Field(default=None, ge=0, le=86400)
    distance_m: float | None = Field(default=None, ge=0, le=99999)
    status: SetStatus = SetStatus.COMPLETED

    @model_validator(mode="before")
    @classmethod
    def apply_backward_compat(cls, data: dict) -> dict:  # type: ignore[type-arg]
        """Backward Compatibility: Sets ohne type → weight_reps mit required-Feldern."""
        if isinstance(data, dict) and "type" not in data:
            data["type"] = SetType.WEIGHT_REPS.value
            # Alte Sets haben reps/weight_kg als required — Defaults setzen falls fehlend
            data.setdefault("reps", data.get("reps", 0))
            data.setdefault("weight_kg", data.get("weight_kg", 0))
        return data

    @model_validator(mode="after")
    def validate_fields_for_type(self) -> "SetInput":
        """Validiert dass die für den Set-Typ erforderlichen Felder vorhanden sind."""
        field_config = SET_TYPE_FIELDS[self.type]
        for field_name in field_config["required"]:
            value = getattr(self, field_name)
            if value is None:
                msg = f"Feld '{field_name}' ist erforderlich für Set-Typ '{self.type.value}'"
                raise ValueError(msg)
        return self


class ExerciseInput(BaseModel):
    """Einzelne Übung mit Sätzen."""

    name: str = Field(..., min_length=1, max_length=100, description="Name der Übung")
    category: ExerciseCategory = Field(..., description="Kategorie")
    sets: list[SetInput] = Field(..., min_length=1, description="Mindestens ein Satz")


class StrengthSessionCreate(BaseModel):
    """Request zum Erstellen einer Krafttraining-Session."""

    date: date
    duration_minutes: int = Field(..., ge=1, le=600, description="Dauer in Minuten")
    exercises: list[ExerciseInput] = Field(..., min_length=1, description="Mindestens eine Übung")
    notes: Optional[str] = Field(None, max_length=2000, description="Notizen")
    rpe: Optional[int] = Field(None, ge=1, le=10, description="RPE (Rate of Perceived Exertion)")


# --- Response Models ---


class SetResponse(BaseModel):
    """Satz in der API-Antwort."""

    type: str = "weight_reps"
    reps: Optional[int] = None
    weight_kg: Optional[float] = None
    duration_sec: Optional[int] = None
    distance_m: Optional[float] = None
    status: str


class ExerciseResponse(BaseModel):
    """Übung in der API-Antwort."""

    name: str
    category: str
    sets: list[SetResponse]


class StrengthMetrics(BaseModel):
    """Auto-berechnete Metriken für Krafttraining."""

    total_exercises: int
    total_sets: int
    total_tonnage_kg: float
    completed_sets: int
    rpe: Optional[int] = None


class LastExerciseSets(BaseModel):
    """Letzte Sätze einer Übung (Quick-Add)."""

    exercise_name: str
    category: str
    sets: list[SetResponse]
    session_date: date


class LastCompleteSessionResponse(BaseModel):
    """Letzte vollstaendige Strength-Session (fuer Clone)."""

    id: int
    date: date
    exercises: list[ExerciseResponse]
    total_tonnage_kg: float
    duration_minutes: Optional[int] = None

File: frontend/src/api/strength.ts
import { apiClient } from './client';

// --- Types ---

export type ExerciseCategory = 'push' | 'pull' | 'legs' | 'core' | 'cardio' | 'drills';
export type SetStatus = 'completed' | 'reduced' | 'skipped';

export type SetType =
  | 'weight_reps'
  | 'bodyweight_reps'
  | 'weighted_bodyweight'
  | 'assisted_bodyweight'
  | 'duration'
  | 'weight_duration'
  | 'distance_duration'
  | 'weight_distance';

export const SET_TYPE_LABELS: Record<SetType, string> = {
  weight_reps: 'Gewicht + Wdh.',
  bodyweight_reps: 'Nur Wdh.',
  weighted_bodyweight: 'Zusatzgewicht',
  assisted_bodyweight: 'Assistiert',
  duration: 'Dauer',
  weight_duration: 'Gewicht + Dauer',
  distance_duration: 'Distanz + Dauer',
  weight_distance: 'Gewicht + Distanz',
};

export const SET_TYPE_OPTIONS: { value: SetType; label: string }[] = [
  { value: 'weight_reps', label: 'Gewicht + Wdh.' },
  { value: 'bodyweight_reps', label: 'Nur Wdh.' },
  { value: 'weighted_bodyweight', label: 'Zusatzgewicht + Wdh.' },
  { value: 'assisted_bodyweight', label: 'Assistiert + Wdh.' },
  { value: 'duration', label: 'Dauer' },
  { value: 'weight_duration', label: 'Gewicht + Dauer' },
  { value: 'distance_duration', label: 'Distanz (+ Dauer)' },
  { value: 'weight_distance', label: 'Gewicht + Distanz' },
];

/** Welche Felder pro Set-Typ angezeigt werden. */
export const SET_TYPE_FIELDS: Record<
  SetType,
  { reps: boolean; weight: boolean; duration: boolean; distance: boolean }
> = {
  weight_reps: { reps: true, weight: true, duration: false, distance: false },
  bodyweight_reps: { reps: true, weight: false, duration: false, distance: false },
  weighted_bodyweight: { reps: true, weight: true, duration: false, distance: false },
  assisted_bodyweight: { reps: true, weight: true, duration: false, distance: false },
  duration: { reps: false, weight: false, duration: true, distance: false },
  weight_duration: { reps: false, weight: true, duration: true, distance: false },
  distance_duration: { reps: false, weight: false, duration: true, distance: true },
  weight_distance: { reps: false, weight: true, duration: false, distance: true },
};

/** Typ-Gruppen für Metriken und Charts. */
export const WEIGHTED_TYPES: SetType[] = [
  'weight_reps',
  'weighted_bodyweight',
  'assisted_bodyweight',
  'weight_duration',
  'weight_distance',
];
export const REP_BASED_TYPES: SetType[] = ['bodyweight_reps'];
export const DURATION_TYPES: SetType[] = ['duration', 'weight_duration', 'distance_duration'];
export const DISTANCE_TYPES: SetType[] = ['distance_duration', 'weight_distance'];

export function isWeightedType(t: string): boolean {
  return WEIGHTED_TYPES.includes(t as SetType);
}
export function isRepBasedType(t: string): boolean {
  return REP_BASED_TYPES.includes(t as SetType);
}
export function isDurationType(t: string): boolean {
  return DURATION_TYPES.includes(t as SetType);
}
export function isDistanceType(t: string): boolean {
  return DISTANCE_TYPES.includes(t as SetType);
}

export interface SetInput {
  type?: SetType;
  reps?: number;
  weight_kg?: number;
  duration_sec?: number;
  distance_m?: number;
  status: SetStatus;
}

export interface ExerciseInput {
  name: string;
  category: ExerciseCategory;
  sets: SetInput[];
}

export interface StrengthSessionCreateParams {
  date: string; // YYYY-MM-DD
  duration_minutes: number;
  exercises: ExerciseInput[];
  notes?: string;
  rpe?: number;
  trainingFile?: File;
  plannedEntryId?: number;
}

export interface StrengthSessionCreateResponse {
  success: boolean;
  session_id: number;
  metrics: {
    total_exercises: number;
    total_sets: number;
    total_tonnage_kg: number;
    total_reps: number;
    total_duration_sec: number;
    total_distance_m: number;
    completed_sets: number;
  };
  file_data?: {
    has_file: boolean;
    hr_avg: number | null;
    hr_max: number | null;
    duration_sec: number | null;
  };
}

export interface LastCompleteSession {
  id: number;
  date: string;
  exercises: ExerciseInput[];
  total_tonnage_kg: number;
  duration_minutes: number | null;
}

export interface LastCompleteSessionResponse {
  found: boolean;
  session: LastCompleteSession | null;
}

export interface LastExerciseResponse {
  found: boolean;
  exercise: {
    exercise_name: string;
    category: string;
    sets: Array<{
      type?: string;
      reps?: number;
      weight_kg?: number;
      duration_sec?: number;
      distance_m?: number;
      status: string;
    }>;
    session_date: string;
  } | null;
}

// --- API Functions ---

export async function createStrengthSession(
  params: StrengthSessionCreateParams,
): Promise<StrengthSessionCreateResponse> {
  const formData = new FormData();
  formData.append('exercises_json', JSON.stringify(params.exercises));
  formData.append('training_date', params.date);
  formData.append('duration_minutes', String(params.duration_minutes));

  if (params.notes) {
    formData.append('notes', params.notes);
  }
  if (params.rpe != null) {
    formData.append('rpe', String(params.rpe));
  }
  if (params.trainingFile) {
    formData.append('training_file', params.trainingFile);
  }
  if (params.plannedEntryId != null) {
    formData.append('planned_entry_id', String(params.plannedEntryId));
  }

  const response = await apiClient.post<StrengthSessionCreateResponse>(
    '/api/v1/sessions/strength',
    formData,
    { headers: { 'Content-Type': 'multipart/form-data' } },
  );
  return response.data;
}

export async function getLastCompleteStrengthSession(): Promise<LastCompleteSessionResponse> {
  const response = await apiClient.get<LastCompleteSessionResponse>(
    '/api/v1/sessions/strength/last-complete',
  );
  return response.data;
}

export interface UpdateExercisesResponse {
  success: boolean;
  session_id: number;
  exercises: Array<{
    name: string;
    category: string;
    sets: Array<{
      type?: string;
      reps?: number;
      weight_kg?: number;
      duration_sec?: number;
      distance_m?: number;
      status: string;
    }>;
  }>;
  metrics: {
    total_exercises: number;
    total_sets: number;
    total_tonnage_kg: number;
    total_reps: number;
    total_duration_sec: number;
    total_distance_m: number;
    completed_sets: number;
  };
}

export async function updateStrengthExercises(
  sessionId: number,
  exercises: ExerciseInput[],
): Promise<UpdateExercisesResponse> {
  const formData = new FormData();
  formData.append('exercises_json', JSON.stringify(exercises));
  const response = await apiClient.patch<UpdateExercisesResponse>(
    `/api/v1/sessions/strength/${sessionId}/exercises`,
    formData,
    { headers: { 'Content-Type': 'multipart/form-data' } },
  );
  return response.data;
}

export async function getLastExerciseSets(exerciseName: string): Promise<LastExerciseResponse> {
  const response = await apiClient.get<LastExerciseResponse>(
    '/api/v1/sessions/strength/last-exercises',
    { params: { exercise_name: exerciseName } },
  );
  return response.data;
}
